### DISCUSSION TITLE
Lowering Asyncify overhead

### DISCUSSION THREAD

**Discussion by iliakan** [WebAssembly/binaryen#5495] [Q&A]
Hello,

Thank you for the great project!

Did I get it right that Asyncify must instrument all WASM functions, constituting the call chain from the top to the one that needs to work asynchronously? 

Only these functions, or anything else?

I just tried to build `emcc -sASYNCIFY`, and the `.wasm` size grew about +70%, kind of too much, so figuring out how to lessen the impact.

**Comment by kripken** [comment]
Looks like a duplicate of https://github.com/WebAssembly/binaryen/discussions/5494 ? Let's continue the discussion there in one place.

**Reply by iliakan** [reply]
Thank you for the response. I closed the issue, because maybe I'm just misunderstanding the situation.

Hopefully, after our discussion I'll be able to send a PR to docs, to make them easier to understand (are they in github as well?).

What is an indirect call? 
Is it call by a pointer, such as `cmd->foo(a, b)`, where `foo` is a function pointer `int (*foo)(int, int)`?
Or something else?

**Reply by kripken** [reply]
Yes, an indirect call in C is a call by a function pointer. In C++ it can also be a virtual function call in a class. The problem is that virtual calls can go to many possible places, and where is determined only at runtime, so the asyncify analysis must include all the possible places in code to instrument).

**Reply by iliakan** [reply]
So when ASYNCIFY comes across an indirect call... Does it instrument all the functions with the matching signature? Or just all the functions? ;)

**Reply by kripken** [reply]
I think right now it assumes any indirect call can go anywhere, which could be improved to just the functions who have a reference taken, and of the right signature,

https://github.com/WebAssembly/binaryen/blob/cd90ef436e9038331ce52e24db3ead6312426e8b/src/passes/Asyncify.cpp#L616-L625

[JSPI](https://v8.dev/blog/jspi) will avoid all this overhead, however.

**Comment by iliakan** [comment]
I have only one async function `launch_editor_js`.

I tried to use `-sASYNCIFY_IGNORE_INDIRECT`, and again there are many instrumented functions.

Here's the top of the advise list:
```
[asyncify] launch_editor_js is an import that can change the state
[asyncify] __wasi_fd_sync is an import that can change the state
...
```

My function is at the top, but then goes `__wasi_fd_sync` and takes almost all app with it.

Why is `__wasi_fd_sync` is instrumented? 

It's in the DEFAULT_ASYNCIFY_IMPORTS list:

https://github.com/emscripten-core/emscripten/blob/main/emcc.py#L102-L110

Is it possible (and will it be ok) to exclude `__wasi_fd_sync`?

**Reply by kripken** [reply]
`fd_sync` can be an async operation. There is asyncify integration for it here:

https://github.com/emscripten-core/emscripten/blob/4000c3b938a51d5f0950b9ab785fcc9dbdc38d6e/src/library_wasi.js#L406-L417

I think adding it to `ASYNCIFY_REMOVE` would ignore it, if you know you don't need that. If that doesn't work though then maybe we need to add code to allow overriding the default asyncify imports with `ASYNCIFY_REMOVE`.

**Reply by iliakan** [reply]
Ah, that's a nice touch! But right, it's not always needed.

I tried to disable it with `-sASYNCIFY_REMOVE=__wasi_fd_sync`

That's the error I got:
```js
Fatal: Asyncify removelist contained an imported function name (use the import list for imports): __wasi_fd_sync
```

**Reply by kripken** [reply]
Ok, in that case then emscripten does need a fix to allow removing things from the default list.

**Reply by iliakan** [reply]
Will it be ok to remove it from the list in [emcc.py](https://github.com/emscripten-core/emscripten/blob/main/emcc.py#L102-L110), while keeping `#ifdef ASYNCIFY` in [library_wasi.js](https://github.com/emscripten-core/emscripten/blob/4000c3b938a51d5f0950b9ab785fcc9dbdc38d6e/src/library_wasi.js#L406-L417) ?

**Reply by kripken** [reply]
No, I think it needs to stay in `emcc.py` as it's the right default. `library_wasi` just defines the code, but doesn't annotate it as an asyncify import.

But we can allow the removelist to override the defaults in `emcc.py`. A PR with that would make sense I think.

**Reply by iliakan** [reply]
I tried to remove the `__wasi_fd_sync` item by editing `emcc.py`, just to see how it goes, and got this Fatal message (above).

Where does it come from? Guess there's something else to change?

**Reply by kripken** [reply]
Hmm, perhaps the remove-list cannot contain an import name in general.

We could change that so the remove-list is used in two ways: First, if it is the name of an import, then the import is removed from the list of imports. Second, if it is *not* the name of an import, then it is passed to the list of things to remove in the Asyncify pass. That is, we don't need to inform the Asyncify pass about removing imports - we can remove that in emcc.py
